### DISCUSSION TITLE
Api is getting call even on Filter Mode Change

### DISCUSSION THREAD

**Discussion by ujjavalavirus** [KevinVandy/material-react-table#354] [Q&A]
I am using Material React table and added filter mode option. When ever I am changing the filter mode or selecting the mode first time. API getting call even without a value. How to fix that?

**Comment by KevinVandy** [comment]
This would probably be due to how you are using react-query or the dependency array of your useEffect.

**Reply by ujjavalavirus** [reply]
I have taken the code same from material - react-table portable and enable the mode option. Now when ever i am selecting/choosing the mode of filter. Its Calling the API. Which is making bad user experiences. Could you please help to resolve. 

import React, {
    useCallback,
    useEffect,
    useMemo,
    useRef,
    useState
} from 'react';
import MaterialReactTable from 'material-react-table';
import { Typography } from '@mui/material';
import {
    QueryClient,
    QueryClientProvider,
    useInfiniteQuery
} from '@tanstack/react-query';

const columns = [
    {
        accessorKey: 'firstName',
        header: 'First Name'
    },
    {
        accessorKey: 'lastName',
        header: 'Last Name'
    },
    {
        accessorKey: 'address',
        header: 'Address'
    },
    {
        accessorKey: 'state',
        header: 'State'
    },
    {
        accessorKey: 'phoneNumber',
        header: 'Phone Number'
    }
];

const fetchSize = 25;

const Example = () => {
    const tableContainerRef = useRef(null); // we can get access to the underlying TableContainer element and react to its scroll events
    const virtualizerInstanceRef = useRef(null); // we can get access to the underlying Virtualizer instance and call its scrollToIndex method

    const [columnFilters, setColumnFilters] = useState([]);
    const [globalFilter, setGlobalFilter] = useState();
    const [sorting, setSorting] = useState([]);

    const {
        data, fetchNextPage, isError, isFetching, isLoading
    } = useInfiniteQuery({
        queryKey: ['table-data', columnFilters, globalFilter, sorting],
        queryFn: async ({ pageParam = 0 }) => {
            const url = new URL('URL');
            url.searchParams.set('start', `${pageParam * fetchSize}`);
            url.searchParams.set('size', `${fetchSize}`);
            url.searchParams.set('filters', JSON.stringify(columnFilters ?? []));
            url.searchParams.set('globalFilter', globalFilter ?? '');
            url.searchParams.set('sorting', JSON.stringify(sorting ?? []));

            const response = await fetch(url.href);
            const json = await response.json();
            return json;
        },
        getNextPageParam: (_lastGroup, groups) => groups.length,
        keepPreviousData: true,
        refetchOnWindowFocus: false
    });

    const flatData = useMemo(
        () => data?.pages.flatMap((page) => page.data) ?? [],
        [data]
    );

    const totalDBRowCount = data?.pages?.[0]?.meta?.totalRowCount ?? 0;
    const totalFetched = flatData.length;

    // called on scroll and possibly on mount to fetch more data as the user scrolls and reaches bottom of table
    const fetchMoreOnBottomReached = useCallback(
        (containerRefElement) => {
            if (containerRefElement) {
                const { scrollHeight, scrollTop, clientHeight } = containerRefElement;
                // once the user has scrolled within 400px of the bottom of the table, fetch more data if we can
                if (
                    scrollHeight - scrollTop - clientHeight < 400
            && !isFetching
            && totalFetched < totalDBRowCount
                ) {
                    fetchNextPage();
                }
            }
        },
        [fetchNextPage, isFetching, totalFetched, totalDBRowCount]
    );

    // scroll to top of table when sorting or filters change
    useEffect(() => {
        if (virtualizerInstanceRef.current) {
            virtualizerInstanceRef.current.scrollToIndex(0);
        }
    }, [sorting, columnFilters, globalFilter]);

    // a check on mount to see if the table is already scrolled to the bottom and immediately needs to fetch more data
    useEffect(() => {
        fetchMoreOnBottomReached(tableContainerRef.current);
    }, [fetchMoreOnBottomReached]);

    return (
      <MaterialReactTable
        columns={ columns }
        data={ flatData }
        enablePagination={ false }
        enableRowNumbers
        enableColumnFilterModes
        enableRowVirtualization // optional, but recommended if it is likely going to be more than 100 rows
        manualFiltering
        manualSorting
        muiTableContainerProps={{
            ref: tableContainerRef, // get access to the table container element
            sx: { maxHeight: '600px' }, // give the table a max height
            onScroll: (
                event // add an event listener to the table container element
            ) => fetchMoreOnBottomReached(event.target)
        }}
        muiToolbarAlertBannerProps={
          isError
              ? {
                  color: 'error',
                  children: 'Error loading data'
              }
              : undefined
        }
        onColumnFiltersChange={ setColumnFilters }
        onGlobalFilterChange={ setGlobalFilter }
        onSortingChange={ setSorting }
        renderBottomToolbarCustomActions={ () => (
          <Typography>
            Fetched {totalFetched} of {totalDBRowCount} total rows.
          </Typography>
        ) }
        state={{
            columnFilters,
            globalFilter,
            isLoading,
            showAlertBanner: isError,
            showProgressBars: isFetching,
            sorting
        }}
        rowVirtualizerProps={{ overscan: 4 }}
      />
    );
};

// const queryClient = new QueryClient();

export default Example;

**Reply by KevinVandy** [reply]
ok I see the problem and will probably ship a fix today

**Reply by ujjavalavirus** [reply]
Any update on this issue?.. How I can get fixed version? @KevinVandy

**Reply by KevinVandy** [reply]
It was fixed in v1.8.0. Check out the change log

**Reply by KevinVandy** [reply]
https://www.material-react-table.com/changelog

**Reply by ujjavalavirus** [reply]
Still its not resolved. If I am selecting between. It calling API with having any value in min and max?
and table.resetcolumnFilters() is also not clearing between mode after selecting. Please help to resolve @KevinVandy

**Reply by KevinVandy** [reply]
Hmm, I had left that specific scenario alone because when changing filter variants, the filter has to change from a primitive value to an array. That should be intentional.

**Reply by ujjavalavirus** [reply]
@KevinVandy And resetColumnFilters()? Can we do something for this ?
So that it remove the mode as well before calling API?

**Reply by KevinVandy** [reply]
Feel free to make a pr
